### PR TITLE
[tune] Added WandbLogger

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This PR adds a `WandbLogger` for logging to the weights & biases service.

W/B has a [rudimentary ray integration](https://github.com/wandb/client/blob/master/wandb/ray/__init__.py). However, it currently does not work out of the box with multiple concurrent runs. In contrast, this logger automatically extracts trial names and IDs, allows grouping of runs by parameter values or trainable names, and can receive a list of metrics and parameters that should not be sent to wandb for logging.

This PR includes a test suite for the new logger. In a second PR, we will add a short example for the documentation.

To be discussed:

1) Instead of specifying excludes, should we specify includes? 
2) Should we move the logger to loggers.py?

To do:

1) Include in docs (after deciding on the location)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests

### CODE REVIEW

File: python/ray/tune/examples/cifar10_pytorch.py [hunk 1]
```diff
@@ -178,7 +180,12 @@ def main(num_samples=10, max_num_epochs=10, gpus_per_trial=2):
         "l1": tune.sample_from(lambda _: 2 ** np.random.randint(2, 9)),
         "l2": tune.sample_from(lambda _: 2 ** np.random.randint(2, 9)),
         "lr": tune.loguniform(1e-4, 1e-1),
-        "batch_size": tune.choice([2, 4, 8, 16])
+        "batch_size": tune.choice([2, 4, 8, 16]),
+        "wandb": {
+            "project": "CIFAR10",
+            "api_key": "1f06d7f72cff2b61773efd8d1d151e71b3c39885",
```
Comment (richardliaw) [inline python/ray/tune/examples/cifar10_pytorch.py:186]: probably want to remove your wandb key

File: python/ray/tune/examples/cifar10_pytorch.py [hunk 2]
```diff
@@ -178,7 +180,12 @@ def main(num_samples=10, max_num_epochs=10, gpus_per_trial=2):
...
+            "group_by": ["batch_size"]
```
Comment (richardliaw) [inline python/ray/tune/examples/cifar10_pytorch.py:187]: we don't need this right?

File: python/ray/tune/examples/cifar10_pytorch.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/examples/cifar10_pytorch.py:186]: I committed the CIFAR example by accident, this file shouldn't be included in the PR at all. Thanks for pointing that out, I will remove it later. The API key is now invalid, too.

File: python/ray/tune/examples/cifar10_pytorch.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/examples/cifar10_pytorch.py:187]: Yeah I think it makes configuration easier if we have less options there, so I will remove this later

File: python/ray/tune/integration/wandb.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                "parameter_2": tune.choice([4, 5, 6]),
+                # wandb configuration
+                "wandb": {
+                    "project": "Optimization_Project",
+                    "api_key_file": "/path/to/file",
+                    "log_config": True
+                }
+            },
+            loggers=DEFAULT_LOGGERS + (WandbLogger, ))
+        ```
```
Comment (richardliaw) [inline python/ray/tune/integration/wandb.py:111]:
```suggestion
    .. code-block:: python

        from ray.tune.logger import WandbLogger, DEFAULT_LOGGERS
        tune.run(
            train_fn,
            config={
                # define search space here
                "parameter_1": tune.choice([1, 2, 3]),
                "parameter_2": tune.choice([4, 5, 6]),
                # wandb configuration
                "wandb": {
                    "project": "Optimization_Project",
                    "api_key_file": "/path/to/file",
                    "log_config": True
                }
            },
            loggers=DEFAULT_LOGGERS + (WandbLogger, ))
        ```
```

**Review by richardliaw — APPROVED** [review]
Looks good; can you make sure the tests run in CI?

File: doc/source/tune/_tutorials/tune-wandb.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
+.. _tune-wandb:
```
Comment (anabranch) [inline doc/source/tune/_tutorials/tune-wandb.rst:1]: LGTM

**Review by bllchmbrs — APPROVED** [review]

File: python/ray/tune/integration/wandb.py [hunk 5]
```diff
@@ -201,3 +216,130 @@ def on_result(self, result):
...
+
+
+    Wandb configuration is done by passing a ``wandb`` key to
+    the ``config`` parameter of ``tune.run()`` (see example below).
+
+    The content of the ``wandb`` config entry is passed to ``wandb.init()``
+    as keyword arguments. The exception are the following settings, which
+    are used to configure the ``WandbTrainableMixin`` itself:
+
+    Args:
```
Comment (richardliaw) [inline python/ray/tune/integration/wandb.py:248]: Shouldn't this documentation go onto `wandb_mixin`?

File: python/ray/tune/examples/wandb_example.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+import numpy as np
```
Comment (richardliaw) [inline python/ray/tune/examples/wandb_example.py:1]: we should include this in BUILD

Comment (richardliaw) [reply python/ray/tune/examples/wandb_example.py:1]: also, this example doesn't actually show up on the docs anywhere (though this is a NIT, we can fix later)

**Review by richardliaw — APPROVED** [review]

File: python/ray/tune/examples/wandb_example.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import wandb
+
+from ray import tune
+from ray.tune import Trainable
+from ray.tune.integration.wandb import WandbLogger, WandbTrainableMixin, \
+    wandb_mixin
+from ray.tune.logger import DEFAULT_LOGGERS
+
+
+def train_function(config, checkpoint=None):
```
Comment (richardliaw) [inline python/ray/tune/examples/wandb_example.py:15]:
```suggestion
def train_function(config, checkpoint_dir=None):
```

File: python/ray/tune/examples/wandb_example.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            "wandb": {
+                "api_key_file": api_key_file,
+                "project": "Wandb_example"
+            }
+        },
+        loggers=DEFAULT_LOGGERS + (WandbLogger, ))
+
+
+@wandb_mixin
+def decorated_train_function(config, checkpoint=None):
```
Comment (richardliaw) [inline python/ray/tune/examples/wandb_example.py:37]:
```suggestion
def decorated_train_function(config, checkpoint_dir=None):
```

**Comment by kisuke95** [comment]
Hi, can you fix these python lint error (https://travis-ci.com/github/ray-project/ray/jobs/366908751#L492) ?

**Comment by krfricke** [comment]
@kisuke95 fixed in #9835

**Pull request merged**